src/modm/platform/core/rp/multicore.hpp

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022, Andrey Kunitsyn
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+#pragma once
+
+#include <modm/platform/device.hpp>
+#include <hardware/structs/sio.h>
+#include <atomic>
+
+namespace modm::platform::multicore {
+	/**
+	 * The RP2040 contains two FIFOs for passing data, messages or ordered events between the two cores. Each FIFO is 32 bits
+	 * wide, and 8 entries deep. One of the FIFOs can only be written by core 0, and read by core 1. The other can only be written
+	 * by core 1, and read by core 0.
+	 */
+	struct Mailbox {
+		static inline bool canRead()
+		{
+			return !!(sio_hw->fifo_st & SIO_FIFO_ST_VLD_BITS);
+		}
+		static inline bool canWrite()
+		{
+			return !!(sio_hw->fifo_st & SIO_FIFO_ST_RDY_BITS);
+		}
+		static inline bool push(uint32_t msg)
+		{
+			if (!canWrite()) return false;
+			sio_hw->fifo_wr = msg;
+			__SEV();
+			return true;
+		}
+		static inline bool pop(uint32_t& msg)
+		{
+			if (!canRead()) return false;
+			msg = sio_hw->fifo_rd;
+			return true;
+		}
+		static inline void pushBlocked(uint32_t msg)
+		{
+			while (!canWrite()) __NOP();
+			sio_hw->fifo_wr = msg;
+			__SEV();
+		}
+		static inline uint32_t popBlocked()
+		{
+			while (!canRead()) __WFE();
+			return sio_hw->fifo_rd;
+		}
+		static void drain()
+		{
+			while (canRead())
+				(void)sio_hw->fifo_rd;
+		}
+	};
+
+	/**
+	 * The RP2040 provides 32 hardware spin locks, which can be used to manage mutually-exclusive access to shared software
+ 	 * and hardware resources.
+ 	 *
+ 	 * Instance 0 allocated for modm usage.
+ 	 *
+ 	 * Acquire a spin lock without disabling interrupts (hence unsafe)
+ 	 */
+	template <uint8_t instance>
+	class SpinLockUnsafe {
+		static_assert(instance < 32);
+		using LockType = volatile uint32_t;
+	private:
+		static inline LockType* getInstance() {
+			return (LockType *) (SIO_BASE + SIO_SPINLOCK0_OFFSET + instance * 4);
+		}
+	public:
+		static void init() {
+			unlock();
+		}
+		static modm_always_inline void lock() {
```

Comment:
```suggestion
		static inline void lock() {
```
(plus everywhere else in the file)
Let's let the compiler decide, I know old modm code still uses this, but it inlines too aggressively.